### DISCUSSION TITLE
Missing basic feature [Impersonation]

### DISCUSSION THREAD

**Discussion by Sipioteo** [zitadel/zitadel#6006] [General]
You have things who are totally unuseful (like storing the phone number with a verification process through a single sms provider) but you have miss the only feature that will make people switch from keycloak to you.

IMPERSONATION.

you know why impersonation is important? Because developers works always with different requirement and impersonation is able to make you fit all.

Customer wants native sms sign in with a custom sms provider? ~> impersonation

Customer wants sign in with a magic code sent through email? ~> impersonation 

Customer wants to make a beautiful and custom sign in/sign up flow with multiple choices and metadata? ~> impersonation

Instead of trying to create a new full apis endpoints to manipulate session, flows and users you should just implement impersonation.

And if you ask yourself why, well the reason it’s that me and all my entrepreneurs colleagues are still using keycloak because you don’t have the basic impersonation feature.

You’re making a great job, just keep it adaptable to all situations or we will continue using keycloak.

Best,
A man who have believed in your project.

**Comment by fforootd** [comment]
Hi 

Thanks for sharing your thoughts, this really helps shaping the future of ZITADEL!

Just to get this right. You would like to have an API where you can claim tokens for any user and impersonate them, right?
So basically Token Exchange as defined in [RFC 8693](https://datatracker.ietf.org/doc/html/rfc8693) or something similar.

> Instead of trying to create a new full apis endpoints to manipulate session, flows and users you should just implement impersonation.

Since the time we started with ZITADEL and today we learned many things on how people would like to interact with it, and we thought it is time to improve our API Design. In doing so we decided to work on a resource based API design that easier exposes all the data stored in ZITADEL. Besides this we wanted to open our authentication capabilities, so that people can build their own Login UIs.

> A man who have believed in your project.

Out of curiosity, don't you believe in this project anymore?
If so, may I ask why?

**Reply by Sipioteo** [reply]
> Just to get this right. You would like to have an API where you can claim tokens for any user and impersonate them, right?
So basically Token Exchange as defined in [RFC 8693](https://datatracker.ietf.org/doc/html/rfc8693) or something similar.

Something like that, or like keycloak, or event a single endpoint where you can request access token, id token and refresh token i t would work.

The session api is a good start, i do the validation and if it work i should be able to request those data and send them to the user so he can use them like normal tokens (refreshing them too with the refresh token). 

Just don't make it too hard to implement, we love keycloak because with two request you can get the data you need without messing with permissions or settings.

> If so, may I ask why?

I'm using Zitadel for a while right now, and i have to admit that when a customer ask me for a solution i always use keycloak. With every upgrade i hope you implement this basic feature but it is always missing and i have to continue using keycloak.
I have one zitadel instance running on my docker, and i will continue following your project in hope this thing will be implemented but i have to do things, and zitadel is not ready to be implemnted in production like projects. At least for now.

**Reply by fforootd** [reply]
I see, thanks for your insights!

@muhlemmer noted that our OIDC lib actually already supports [token exchange](https://github.com/zitadel/oidc/pull/255).
So we would have two options to proceed here:

1. Implement Token Exchange in ZITADEL
2. Create a new method on the new session API that allows impersonation. I.e /impersonation

@muhlemmer @livio-a @hifabienne what is your take on this?

**Reply by Sipioteo** [reply]
News about this one?

**Reply by muhlemmer** [reply]
> @muhlemmer @livio-a @hifabienne what is your take on

Sounds doable to me.

**Reply by muhlemmer** [reply]
> News about this one?

We inplement features prioritized on customer demand. Based on that we already have a full backlog and roadmap.

For all new features any of the following conditions should be met:

1. A request from a paying customer.
2. Popular demand from community users.
3. A contribution/ PR.

As far as i can see regarding (2), there was only feedback about this feature missing and it being the primary reason for the OP to abandon the project. If there are any ACTIVE users left that need this feature, I invite them  to make themselves known.

Please refrain from pinging on features, because it will not motivate us to do it faster.

**Comment by matthewtusker** [comment]
Hello, we would be interested in impersonation. We're trying to replace an authentication system and impersonation is our current road block. We have a possible solution that we're pursuing, but it requires a lot of working around Zitadel so it doesn't feel just quite right. If this was implemented in Zitadel using RFC 8693 we as a team would be much more comfortable. Everything else that Zitadel provides for our system has been great, and impersonation would be the icing on the cake.

**Reply by fforootd** [reply]
Thanks for letting us know, while we do not have an ETA for this we clearly hear the rising demand.

**Comment by bhaugeea** [comment]
I would like to second matthew above. I'll add that while it's debatable whether impersonation in production is a good idea, development is a different story. That's where it's non-negotiable for our use case.

**Comment by hreddy-klaviyo** [comment]
We're doing exploratory analysis on Zitadel and this is one of the critical missing features for our company. We have to agree with the above comments about impersonation feature set.

Thanks for the great work, the existing toolset looks awesome

**Comment by Royallyroyally** [comment]
Hi, great product! We are currently using another auth supplier but does not meet the same specs as Zitadel, so we are looking to move.  However we do require impersonation so that our partners can view their end clients screens and offer support. I would be great to find out when / if this is coming. 

Thanks,

**Comment by hifabienne** [comment]
We are currently investing in figuring out what needs to implemented in ZITADEL to provide impersonation.
Can't give you a timeline yet. But will keep you posted and link the issues as soon as they are ready.

**Reply by dm17** [reply]
Please keep us posted, thanks!

**Comment by HungryHowies** [comment]
+1 on  impersonation

**Comment by jacob-buckaroo** [comment]
For us this feature is also critical to have.

**Reply by Sipioteo** [reply]
> For us this feature is also critical to have. 

It's there. They implemented it very well

**Reply by jacob-buckaroo** [reply]
Can you provide a link to the feature you're referring to?

**Reply by muhlemmer** [reply]
No, impersonation is not yet implemented in zitadel.

**Reply by jacob-buckaroo** [reply]
Thanks for the update @muhlemmer.

**Reply by mffap** [reply]
In understand that for this use case impersonation must solve the following: 
- Be able to access a client app as if / on-behalf of a user
- Audit log must show that all actions were executed by the impersonating user on-behalf of the impersonated user

**Reply by dm17** [reply]
@muhlemmer it seems like a point of competition... None of the nice modern auth providers except for SuperTokens seem to support impersonation natively. I'm hesitant to go with SuperTokens for other reasons. Ory seemed like the best, but it also does not support impersonation natively. I'd go with Zitadel if it had a well documented impersionation guide.

**Comment by jhass** [comment]
You should be able to get an access token for any user using the session and auth request APIs, that is just creating a session for the desired user, creating an auth request in the backend and validating the auth request with the created session:

```mermaid
sequenceDiagram
    Client->>+ImpersonationService: Impersonate($loginName, <br>$clientId,<br> $redirectUri,<br>$codeChallenge=Base64URLEncode(SHA256($codeVerifier=randomString)))
    ImpersonationService->>+Zitadel: POST /oauth/v2/token($serviceUserCredentials)
    Zitadel->>-ImpersonationService: $serviceUserAccessToken
    ImpersonationService->>+Zitadel: GET /oauth/v2/authorize?clientId=$clientId<br>&redirectUri=$redirectUri<br>&response_type=code<br>&scope=openid%20email%20offline_access<br>&code_challenge=$codeChallenge<br>&code_challenge_method=S256<br>Headers: x-zitadel-login-client: $serviceUserId
    Zitadel->>-ImpersonationService: Location: /login?authRequest=$authRequestId
    ImpersonationService->>+Zitadel: POST /v2beta/sessions<br>Body: {"checks":{"user":{"loginName": $loginName}}}<br>Headers: Authorization: Bearer $serviceUserAccessToken
    Zitadel->>-ImpersonationService: $sessionId, $sessionToken
    ImpersonationService->>+Zitadel: POST /v2beta/oidc/auth_requests/$authRequestId<br>Body: {"session":{"sessionId": $sessionId, "sessionToken": $sessionToken}}<br>Headers: Authorization: Bearer $serviceUserAccessToken
    Zitadel->>-ImpersonationService: {"callbackUrl": $redirectUri?code=$authCode}
    ImpersonationService->>-Client: {"callbackUrl": $redirectUri?code=$authCode}
    Client->>+Zitadel: POST /oauth/v2/token<br>Body: grant_type=authorization_code<br>&code=$authCode<br>&client_id=$clientId<br>&redirect_uri=$redirectUri<br>&code_verifier=$codeVerifier
    Zitadel->>-Client: $accessToken, $idToken, ...
```

I guess you might also just create an additional app that the impersonation service uses to fetch the user access token itself, instead of proxying the client id, code challenge and redirect URI from the client.

What this doesn't really provide is auditing.

**Reply by tusharmctrl** [reply]
Hello, @jhass. Thank you for the detailed explanation. I've been attempting to accomplish the same thing, and it's giving me this issue. Do you have any idea what I'm doing wrong? A little help would be greatly appreciated. Many thanks.

- Here is my code snippet and error I am facing.

```
const axios = require("axios");
try {
  axios
    .get("https://my-zitadel-cloud.com/oauth/v2/authorize?client_id=myClientId&response_type=code&scope=openid email profile urn:zitadel:iam:org:project:id:zitadel:aud openid&code_challenge=dZtdUQeFnaFhUOBYKdSkJ3iiZRRHXREH1svXez1WZQ4&code_challenge_method=S256&redirect_uri=http://localhost:3000/api/auth/callback/zitadel")
    .then((data) => {
      const requestId = data.request.path.split("=")[1];
      axios
        .post(
          "https://my-zitadel-cloud.com/v2beta/sessions",
          {
            checks: {
              user: { loginName: "username" },
            },
          },
          {
            headers: {
              Authorization:
                "Bearer myServiceUserAccessToken",
            },
          }
        )
        .then((data) => {
            console.log(data);
          const sessionId = data.data.sessionId;
          const sessionToken = data.data.sessionToken;
          axios
            .post(
              `https://my-zitadel-cloud.com/v2beta/oidc/auth_requests/${requestId}`,
              {
                session: {
                  sessionId: sessionId,
                  sessionToken: sessionToken,
                },
              },
              {
                headers: {
                  Authorization:
                    "Bearer myServiceUserAccessToken",
                },
              }
            )
            .then((data) => console.log(data))
            .catch((e) => console.log(e));
        })
        .catch((e) => console.log(e));
    })
    .catch((err) => console.log(err));
} catch (error) {
  console.log(error);
}

```

![image](https://github.com/zitadel/zitadel/assets/141150938/ab034bc4-c8fd-4c4b-9d96-86573065a290)

**Reply by jhass** [reply]
Hey @tusharmctrl,

It is crucial to include the `x-zitadel-login-client` header in the request to `/oauth/v2/authorize`, its value should be your authenticating service user's ID or if you're authenticating your service using an API app that app's client ID. Including this header is what instructs Zitadel to create an authRequest of the new kind that's available and controllable via the APIs. You'll notice the location header in the response will change its format slightly, from `/ui/login/login?authRequestID=...` to `/login?authRequest=V2_....`. It's the latter format of ID (including the `V2_` prefix) that you want and need.

```js
axios
    .get(..., 
       headers: { 'x-zitadel-login-client': '$serviceUserID' })
    .then((data) => {
```

**Reply by tusharmctrl** [reply]
Many Many thanks. Much appreciated.

**Reply by tusharmctrl** [reply]
Hi @jhass, I hope you're doing good. One more problem I have been facing is - when I successfully get the token from zitadel, I am successfully able to `introspect` it. It gives me all the details of the token. But funny thing is - when I try to use that token in any API, such as - 

`https://my-company.zitadel.cloud/auth/v1/users/me`
Auth Token in Headers Authorization: Bearer `token`

It throws this error:
![image](https://github.com/zitadel/zitadel/assets/141150938/f38f692a-82d4-4944-b409-d7883749d3ce)

I dont think there's a problem with my code, but here's the snippet.

``` 
userImpersonate: protectedProcedure.input(z.object({ preferred_username: z.string() })).mutation(async ({ input }) => {
    const authorizeUrl = `${env.NEXTAUTH_ZITADEL_ENDPOINT}/oauth/v2/authorize`;
    const sessionsUrl = `${env.NEXTAUTH_ZITADEL_ENDPOINT}/v2beta/sessions`;
    const codeVerifier = "3239b6c04054ee8197572bb44153e641b6da6cf4653439c525f03aec4ea52c99";
    const codeChallenge = "Uu5QDln6-4FwfXIQThSItmnHAZ1tmSkhI4tb-3x1OiE";
    const axiosInstance = axios.create();
    axiosInstance.defaults.maxRedirects = 0;
    axiosInstance.interceptors.response.use(
      (response) => response,
      async (error) => {
        const axiosError = error as AxiosError;
        if (axiosError.response && [301, 302].includes(axiosError.response.status)) {
          const redirectUrl = axiosError.response;
          return Promise.resolve(redirectUrl);
        }
        return Promise.reject(error);
      }
    );
    try {
      const authorizeResponse = await axiosInstance.get<string>(`${authorizeUrl}`, {
        params: {
          client_id: env.NEXTAUTH_ZITADEL_CLIENT_ID,
          response_type: "code",
          scope: "openid email profile urn:zitadel:iam:org:project:id:zitadel:aud openid",
          code_challenge: codeChallenge,
          code_challenge_method: "S256",
          redirect_uri: "http://localhost:3000/api/auth/callback/zitadel"
        },
        headers: {
          "x-zitadel-login-client": env.NEXTAUTH_ZITADEL_SERVICE_USER_ID
        }
      });
      const authRequestValue = (authorizeResponse.data.match(/authRequest=([^\s]+)\">Found<\/a>/) || [])[1];
      if (authRequestValue) {
        const getAuthCodeUrl = `${env.NEXTAUTH_ZITADEL_ENDPOINT}/v2beta/oidc/auth_requests/${authRequestValue}`;
        const getSessionIdAndToken = await axiosInstance.post<ImpersonateSessionResponse>(
          sessionsUrl,
          {
            checks: {
              user: {
                loginName: input.preferred_username
              }
            }
          },
          {
            headers: {
              Authorization: `Bearer ${env.NEXTAUTH_ZITADEL_SERVICE_USER_TOKEN}`
            }
          }
        );
        const { sessionId, sessionToken } = getSessionIdAndToken.data;
        const getAuthCodeResponse = await axiosInstance.post<ImpersonateAuthCodeResponse>(
          getAuthCodeUrl,
          {
            session: {
              sessionId,
              sessionToken
            }
          },
          {
            headers: {
              Authorization: `Bearer ${env.NEXTAUTH_ZITADEL_SERVICE_USER_TOKEN}`
            }
          }
        );
        const getCodeUrl = getAuthCodeResponse.data.callbackUrl;
        const urlPms = new URLSearchParams(getCodeUrl.split("?")[1]);
        const authCode = urlPms.get("code");
        const getTokenUrl = `${env.NEXTAUTH_ZITADEL_ENDPOINT}/oauth/v2/token`;
        const getTokenResponse = await axiosInstance.post<ImpersonateTokenResponse>(
          getTokenUrl,
          {},
          {
            params: {
              grant_type: "authorization_code",
              code: authCode,
              client_id: env.NEXTAUTH_ZITADEL_CLIENT_ID,
              redirect_uri: "http://localhost:3000/api/auth/callback/zitadel",
              code_verifier: codeVerifier,
              client_secret: env.NEXTAUTH_ZITADEL_CLIENT_SECRET
            },
            headers: {
              "x-zitadel-login-client": env.NEXTAUTH_ZITADEL_SERVICE_USER_ID
            }
          }
        );
        const { access_token, expires_in, id_token, token_type } = getTokenResponse.data;
        return {
          data: { access_token, expires_in, id_token, token_type },
          code: 200,
          status: "success",
          message: "Token fetched successfully"
        };
      }
    } catch (error) {
      console.log(error);
      return {
        data: null,
        code: (error as { [key: string]: { [key: string]: string } })?.response?.status ?? 400,
        status: "error",
        message: (error as { [key: string]: string })?.[`message`] ?? "Something went wrong. Try later."
      };
    }
  })
```

Here is the snippet of my code. Thank you in advance. Appreciate it.

**Reply by jhass** [reply]
I have to admit I never tried any other endpoints beyond token verification for this, as my usecases would be just getting a JWT that validates against Zitadel's JWK and has the necessary user data. 

There seem to be additional checks on the token when querying endpoints just as the users/me resource. First of all it seems to require that the token was created with the `urn:zitadel:iam:org:project:id:zitadel:aud` scope, then it requires any kind of authentication to actually have happened. An easy way to automatically do this, if your requirements permit it, is to enable the email OTP MFA method for the user, include `"challenges": {"otpEmail": {"returnCode": {}}` when creating the session and immediately validating the session with an update session call using the returned code. However doing this in a default configuration of Zitadel with force MFA disabled I still get an "mfa required" error on the users/me resource :(

**Comment by ilterugur** [comment]
I would love to use zitadel but I need to use it with custom SMS providers. The impersonation feature would allow us to cover lots of scenarios like that. I'm pretty sure there are lots of people who need custom solutions (just like I've mentioned), figure out zitadel doesn't support impersonation but don't mention it in discussions and consider using another auth library instead of zitadel.

Currently, I'm researching open-source robust authentication technologies and zitadel is currently my favourite by far. Still, I don't think I would be able to use this project on most of my projects because of this missing feature.

**Reply by dm17** [reply]
Indeed, I don't understand why there are so many startups in this space and almost all refuse to implement features that people want and that keycloak has... It would be easy for them to differentiate themselves. Top ones engineering-wise seem to be Ory, Authentik and Zitadel.

**Comment by hifabienne** [comment]
Hello everyone.
We have had a look at how to implement impersonations, there are two different use cases we see.
1. Supporter wants to impersonate user
2. You want to exchange your token from an external service for a ZITADEL token

One question that is still open, who is allowed to do impersonation.
What would you expect? Some ideas we had:
- Enable impersonation on the instance/organization --> users of that organization are allowed to do impersonation
- Add a specific manager role e.g. IMPERSONATOR, that is allowed to do impersonation.

Would that be enough? Or would you expect something else.

You can track all the issues for the different usecases of impersonation here: https://github.com/zitadel/zitadel/issues/7208

**Reply by hreddy-klaviyo** [reply]
In our use case, we would love option 1 (Supporter wants to impersonate user). I've seen the usecase 2 but less prevalent.

re: impersonation expectations ->
I like the idea and would like to share how it works right now in our company:
Customer Service agents belong to the Service Provider Org. This means a few steps on our end
1. They gather explicit consent from the customer for accessing their org (limited policies) and a user's sessions.
2. They impersonate into the user's existing session on the service
3. Once the interaction is complete, then the role is revoked for the customer service agent from that org
 
 I feel the manager role you mentioned works really well with some manual lift on our end
![image](https://github.com/zitadel/zitadel/assets/139139503/11ee17aa-ca86-4d34-9a44-56a8658737bc)

The complications
I would like to hear more how impersonation would work with use cases like the below (Authorize users outside your organization) which are very common in our company
https://zitadel.com/blog/multi-tenancy-with-organizations

![image](https://github.com/zitadel/zitadel/assets/139139503/25fc1659-7884-440f-a5a5-b9f47b865bd5)

**Reply by hifabienne** [reply]
Thanks for your description.
The manager roles can be given on different organizations.
So this could also work if the user is from a different organization.

**Reply by ilterugur** [reply]
In my case, I need to use a country-local SMS service provider.

I've always developed my own auth on my previous projects so I haven't started to use zitadel yet and I don't know if it is a stupid suggestion (or if it would cause a security leak) but something like a special API endpoint which gives me jwt tokens would be enough in my case.

Example scenario:
1- User tries to log in using phone number in frontend. This sends their phone number in a request to our backend server
2- Our backend server sends an SMS including the verification code to the phone number that user sent in the request
3- User sends the verification code (also with the phone number related to it) in the request payload to our backend server using another endpoint
4- Backend server checks if that one-time usage code matches with user's phone number
5- If it matches, backend server forwards that phone number to zitadel using the new impersonation endpoint
6- zitadel finds user by their phone number and returns jwt tokens
7- Backend server forwards that tokens to the frontend
8- Frontend uses these tokens on future requests. From this step, everything works the same way like we didn't do anything related to impersonation

I think I can handle this case somehow without using zitadel by connecting my own backend to zitadel's database; while still using zitadel on other login options. But I use MongoDB and I'm not sure if it's reasonable to install a database adapter to my backend just for a specific login operation. I think it isn't the right way. So I think an impersonation endpoint would be useful to cover cases like this one. We can send some login provider-related unique information (email, phone number etc.) to zitadel using my backend server, zitadel could find related user and return jwt tokens & user id to my backend server. I mean my backend server can act as a proxy between frontend and zitadel's user management system to solve this case

**Reply by fforootd** [reply]
@ilterugur I guess you could achieve this by using our [session-api ](https://zitadel.com/docs/apis/resources/session_service/session-service-create-session) by just creating a new session with a username only 😁 after you did your verifications. This should return you a session with a session_token.

**Comment by metafeather** [comment]
I recently used to PropelAuth at a company and they have a very user friendly User Impersonation implementation with a UI and audit trail, these docs basically outline what I would like Zitadel to have:
- https://docs.propelauth.com/overview/user-management/user-impersonation
